src/tree/hist/evaluate_splits.h

```diff
@@ -91,49 +96,118 @@ template <typename GradientSumT, typename ExpandEntry> class HistEvaluator {
       iend = static_cast<int32_t>(cut_ptr[fidx]) - 1;
     }
 
+    auto calc_bin_value = [&](auto i) {
```

Comment:
This can be split into multiple functions, but then we have have lots of duplicated code.